apis/crd/v1alpha1/cloudentityselector_webhook.go

```diff
@@ -218,28 +218,123 @@ func (r *CloudEntitySelector) validateMatchSections() error {
 	// In AWS, Vpc name(in vpcMatch section) with either vm id or vm name(in vmMatch section) is not supported
 	if cloudProviderType == AzureCloudProvider {
 		for _, m := range r.Spec.VMSelector {
-			if m.VpcMatch != nil {
-				if len(strings.TrimSpace(m.VpcMatch.MatchName)) != 0 {
-					return fmt.Errorf("matchName is not supported in vpcMatch, " +
-						"use matchID instead of matchName")
-				}
+			if m.VpcMatch != nil && len(strings.TrimSpace(m.VpcMatch.MatchName)) != 0 {
+				return fmt.Errorf("matchName is not supported in vpcMatch, " +
+					"use matchID instead of matchName")
 			}
 		}
 	} else {
 		for _, m := range r.Spec.VMSelector {
-			if m.VpcMatch != nil {
-				if len(strings.TrimSpace(m.VpcMatch.MatchName)) != 0 {
-					for _, vmMatch := range m.VMMatch {
-						if len(strings.TrimSpace(vmMatch.MatchID)) != 0 ||
-							len(strings.TrimSpace(vmMatch.MatchName)) != 0 {
-							return fmt.Errorf("vpc matchName with either vm matchID" +
-								" or vm matchName is not supported, use vpc matchID instead" +
-								" of vpc matchName")
-						}
+			if m.VpcMatch != nil && len(strings.TrimSpace(m.VpcMatch.MatchName)) != 0 {
+				for _, vmMatch := range m.VMMatch {
+					if len(strings.TrimSpace(vmMatch.MatchID)) != 0 ||
+						len(strings.TrimSpace(vmMatch.MatchName)) != 0 {
+						return fmt.Errorf("vpc matchName with either vm matchID" +
+							" or vm matchName is not supported, use vpc matchID instead" +
+							" of vpc matchName")
 					}
 				}
 			}
 		}
 	}
+
+	// Block ambiguous match combinations as it can result in unpredictable behavior wrt "agented" configuration.
+	// vpcIDOnlyMatch map holds VPC ID as key if the selector contains only vpcMatch matchID.
+	// vmIDOnlyMatch map holds VM ID as key if the selector contains only vmMatch matchID.
+	// vmNameOnlyMatch map holds VM Name as key if the selector contains only vmMatch matchName.
+	// vmIDWithVpcMatch map holds VM ID and VPC ID as key if the selector contains vpcMatch matchID and vmMatch matchID.
+	// vmNameWithVpcMatch map holds VM Name and VPC ID as key if the selector contains vpcMatch matchID and vmMatch matchName.
+	vpcIDOnlyMatch := make(map[string]struct{})
+	vmIDOnlyMatch := make(map[string]struct{})
+	vmNameOnlyMatch := make(map[string]struct{})
+	vmIDWithVpcMatch := make(map[string]struct{})
+	vmNameWithVpcMatch := make(map[string]struct{})
+
+	for _, m := range r.Spec.VMSelector {
```

Comment:
This for block can be re-written as:

```suggestion
	var err error
	for _, m := range r.Spec.VMSelector {
		if m.VpcMatch != nil {
			err = r.validateVPCMatch(m, vmIDOnlyMatch, vpcIDOnlyMatch, vmIDWithVpcMatch, vmNameWithVpcMatch)
		} else {
			err = r.validateVMMatch(m, vmIDOnlyMatch, vmNameOnlyMatch, vmIDWithVpcMatch)
		}

		if err != nil {
			return err
		}
	}
```

Reply:
done, bit had to change the variable name while declaring as its already used before.

Reply:
I actually meant to check the err outside the if else block, as it is repetitive.
if err != nil {
			return err
		}